returnn/tf/util/data.py

```diff
@@ -1088,25 +1099,42 @@ def get_default_new_axis_for_dim_tag(self, dim_tag):
     else:
       return self.batch_ndim  # add it at the end
 
-  def copy_add_dim_by_tag(self, dim_tag, unbroadcast=False, axis=None):
+  def copy_add_dim_by_tag(self, dim_tag, unbroadcast=False, axis=None, dyn_size=None):
     """
     :param DimensionTag dim_tag:
     :param bool unbroadcast:
     :param int|None axis:
+    :param tf.Tensor|int|None dyn_size:
+      For unbroadcast, if we do not want to rely on dim_tag.dimension.
     :rtype: Data
     """
     if axis is None:
       axis = self.get_default_new_axis_for_dim_tag(dim_tag=dim_tag)
+    if dyn_size is None:
+      dyn_size = dim_tag.dimension
+
+    def maybe_unbroadcast_placeholder(data):
```

Comment:
You don't need the `data` argument. You can directly access `res`.
Try to keep the code as simple as possible.